### PR TITLE
Mismatch issue of Documents shown and Citation number in text fix

### PR DESCRIPTION
**Pull request by joachim-danswer**
When document order presented to LLM differs from order shown to user, wrong doc numbers are cited.

Fix:
 - SearchTool.get_search_result  returns now final and initial ranking
 - initial ranking is passed through a few objects and used for replacement in citation processing

Notes:
 - the citation_num in the CitationInfo() object has not been changed.

## Description
[Provide a brief description of the changes in this PR]

## How Has This Been Tested?
[Describe the tests you ran to verify your changes]

## Accepted Risk (provide if relevant)
N/A

## Related Issue(s) (provide if relevant)
N/A

## Mental Checklist:
- All of the automated tests pass
- All PR comments are addressed and marked resolved
- If there are migrations, they have been rebased to latest main
- If there are new dependencies, they are added to the requirements
- If there are new environment variables, they are added to all of the deployment methods
- If there are new APIs that don't require auth, they are added to PUBLIC_ENDPOINT_SPECS
- Docker images build and basic functionalities work
- Author has done a final read through of the PR right before merge

## Backporting (check the box to trigger backport action)
Note: You have to check that the action passes, otherwise resolve the conflicts manually and tag the patches.
- [ ] This PR should be backported (make sure to check that the backport attempt succeeds)

### CODE REVIEW

File: backend/danswer/chat/stream_processing/answer_response_handler.py [hunk 1]
```diff
@@ -35,13 +35,18 @@ def handle_response_part(
 
 class CitationResponseHandler(AnswerResponseHandler):
     def __init__(
-        self, context_docs: list[LlmDoc], doc_id_to_rank_map: DocumentIdOrderMapping
+        self,
+        context_docs: list[LlmDoc],
+        doc_id_to_rank_map: DocumentIdOrderMapping,
```
Comment (yuhongsun96) [inline backend/danswer/chat/stream_processing/answer_response_handler.py:40]: Prefer putting the display_doc_order_list within the DocumentIdOrderMapping

File: backend/danswer/chat/answer.py [hunk 2]
```diff
@@ -224,6 +226,7 @@ def _get_response(self, llm_calls: list[LLMCall]) -> AnswerStream:
         answer_handler = CitationResponseHandler(
             context_docs=search_result,
             doc_id_to_rank_map=map_document_id_order(search_result),
```
Comment (yuhongsun96) [inline backend/danswer/chat/answer.py:228]: Would create the single object with both in this function here, so above we can do processed_search_result, displayed_search_results

File: backend/danswer/chat/answer.py [hunk 3]
```diff
@@ -206,7 +206,9 @@ def _get_response(self, llm_calls: list[LLMCall]) -> AnswerStream:
         # + figure out what the next LLM call should be
         tool_call_handler = ToolResponseHandler(current_llm_call.tools)
 
-        search_result = SearchTool.get_search_result(current_llm_call) or []
+        search_result, displayed_search_results_map = SearchTool.get_search_result(
+            current_llm_call
+        ) or ([], {})
```
Comment (yuhongsun96) [inline backend/danswer/chat/answer.py:211]: Pref ([], []) for consistency and building the ordering in the call below similar to the other object. For reducing complexity for future readers

**Review by yuhongsun96 — COMMENTED** [review]
Code best practices

File: backend/danswer/chat/stream_processing/citation_processing.py [hunk 4]
```diff
@@ -22,12 +22,16 @@ def __init__(
         self,
         context_docs: list[LlmDoc],
         doc_id_to_rank_map: DocumentIdOrderMapping,
+        display_doc_order_dict: dict[str, int],
         stop_stream: str | None = STOP_STREAM_PAT,
     ):
         self.context_docs = context_docs
         self.doc_id_to_rank_map = doc_id_to_rank_map
         self.stop_stream = stop_stream
         self.order_mapping = doc_id_to_rank_map.order_mapping
```
Comment (yuhongsun96) [inline backend/danswer/chat/stream_processing/citation_processing.py:31]: This attribute would contain both that way we have a single cohesive "state object" we can pass to the function

**Review by hagen-danswer — APPROVED** [review]
lgtm!!!!!!! @yuhongsun96 fault if anything goes wrong

**Pull request merged**